man/join2.1
.TH  Tukubai 1 "19 Jun 2022" "usp Tukubai" "Tukubai コマンド マニュアル"

.br
join2(1)
.br

.br
【名前】
.br

.br
join2\ :\ トランザクションファイルにマスタファイルを連結(一致しなかった行はダミー
.br
データへ置換)
.br

.br
【書式】
.br

.br
Usage\ \ \ :\ join2\ [-d<string>|+<string>]\ key=<key>\ <master>\ [<tran>]
.br
Version\ :\ Sun\ Jun\ 19\ 23:55:51\ JST\ 2022
.br
Edition\ :\ 1
.br

.br
【説明】
.br

.br
tranのkey=<n>で指定したキーフィールドがmasterの第1フィールド(キーフィールド)と
.br
一致した行をtranから抽出し、masterの情報を連結して出力する。一致しない行はダミ
.br
ーデータ*を連結して出力する。ダミーデータは-dオプションまたは+オプションにて指
.br
定できる。
.br

.br
masterに-を指定すると標準入力をマスタファイルとする。tranが無指定かまたは-が指
.br
定されている場合には、標準入力がトランザクションファイルとなる。
.br

.br
masterの第1フィールドおよびtranの第<n>フィールドは必ず昇順で整列されていること
.br
が条件となる。また、masterのキーフィールド(第1フィールド)はユニークでなければな
.br
らない。(第1フィールドが同じ値をもつ行が複数あってはならない)。tranについてはこ
.br
の制約はなく、キーフィールド(第<n>フィールド)が同じ値の行はいくつあっても構わな
.br
い。
.br

.br
キー指定にひとつでも0があると全フィールドを指定したことになる
.br

.br
tranでキーが整列されていないものはjoin2(1)では処理できない。その場合は一旦tran
.br
を整列してからjoin2(1)で処理するか、join2(1)のかわりにcjoin2(1)を使って処理すれ
.br
ばよい。
.br

.br
【例1】基本パターン
.br

.br

  $ cat master
  0000003 杉山______ 26 F
  0000005 崎村______ 50 F
  0000007 梶川______ 42 F
  0000010 柳本______ 50 F

  $

.br

  $ cat kekka
  0000000 91 59 20 76 54
  0000001 46 39 8  5  21
  0000003 30 50 71 36 30
  0000004 58 71 20 10 6
  0000005 82 79 16 21 80
  0000007 50 2  33 15 62
  0000008 52 91 44 9  0
  0000009 60 89 33 18 6
  0000010 95 60 35 93 76
  0000011 92 56 83 96 75

  $

.br
masterに存在しない行は次のように*で補完して出力される。
.br

.br

  $ join2 key=1 master kekka
  0000000 ********** ** * 91 59 20 76 54
  0000001 ********** ** * 46 39 8  5  21
  0000003 杉山______ 26 F 30 50 71 36 30
  0000004 ********** ** * 58 71 20 10 6
  0000005 崎村______ 50 F 82 79 16 21 80
  0000007 梶川______ 42 F 50 2  33 15 62
  0000008 ********** ** * 52 91 44 9  0
  0000009 ********** ** * 60 89 33 18 6
  0000010 柳本______ 50 F 95 60 35 93 76
  0000011 ********** ** * 92 56 83 96 75

  $

.br
【例2】
.br

.br
左から順に連続した複数のフィールドをキーに指定する場合次のようになる。第2、第3
.br
フィールドでキーが一致するかどうかをチェックしている。
.br

.br

  $ cat master
  A 0000003 杉山______ 26 F
  A 0000005 崎村______ 50 F
  B 0000007 梶川______ 42 F
  C 0000010 柳本______ 50 F

  $

.br

  $ cat kekka
  1 A 0000000 91 59 20 76 54
  2 A 0000001 46 39 8  5  21
  3 A 0000003 30 50 71 36 30
  4 A 0000004 58 71 20 10 6
  5 A 0000005 82 79 16 21 80
  6 B 0000007 50 2  33 15 62
  7 B 0000008 52 91 44 9  0
  8 C 0000009 60 89 33 18 6
  9 C 0000010 95 60 35 93 76
  10 C 0000011 92 56 83 96 75

  $

.br

  $ join2 key=2/3 master kekka
  1 A 0000000 ********** ** * 91 59 20 76 54
  2 A 0000001 ********** ** * 46 39 8  5  21
  3 A 0000003 杉山______ 26 F 30 50 71 36 30
  4 A 0000004 ********** ** * 58 71 20 10 6
  5 A 0000005 崎村______ 50 F 82 79 16 21 80
  6 B 0000007 梶川______ 42 F 50 2  33 15 62
  7 B 0000008 ********** ** * 52 91 44 9  0
  8 C 0000009 ********** ** * 60 89 33 18 6
  9 C 0000010 柳本______ 50 F 95 60 35 93 76
  10 C 0000011 ********** ** * 92 56 83 96 75

  $

.br
【例3】
.br

.br
+<string>はダミーデータとして補完する文字を指定するオプション。+の後に補完で使
.br
用する文字を指定する。
.br

.br

  $ cat master
  0000003 杉山______ 26 F
  0000005 崎村______ 50 F
  0000007 梶川______ 42 F
  0000010 柳本______ 50 F

  $

.br

  $ cat kekka
  0000000 91 59 20 76 54
  0000001 46 39 8  5  21
  0000003 30 50 71 36 30
  0000004 58 71 20 10 6
  0000005 82 79 16 21 80
  0000007 50 2  33 15 62
  0000008 52 91 44 9  0
  0000009 60 89 33 18 6
  0000010 95 60 35 93 76
  0000011 92 56 83 96 75

  $

.br

  $ join2 +@ key=1 master kekka
  0000000 @ @ @ 91 59 20 76 54
  0000001 @ @ @ 46 39 8  5  21
  0000003 杉山______ 26 F 30 50 71 36 30
  0000004 @ @ @ 58 71 20 10 6
  0000005 崎村______ 50 F 82 79 16 21 80
  0000007 梶川______ 42 F 50 2  33 15 62
  0000008 @ @ @ 52 91 44 9  0
  0000009 @ @ @ 60 89 33 18 6
  0000010 柳本______ 50 F 95 60 35 93 76
  0000011 @ @ @ 92 56 83 96 75

  $

.br
【備考】
.br

.br
masterが0バイトの時は何も連結せずtranそのものを出力する。
.br

.br

  $ : > master

  $ join2 key=1 master tran > data

  $ cmp tran data         ←同じデータが出力される

.br
masterのフィールドを保証したい時は、例えば次のようなコードを書く。
.br

.br

  $ [ ! -s master ] && echo 0000000 0 > master

  $ join2 key=1 master tran

.br
【関連項目】
.br

.br
cjoin0(1)、cjoin1(1)、cjoin2(1)、join0(1)、join1(1)、loopj(1)、loopx(1)、up3(1)
.br
、マスタファイル(5)、トランザクションファイル(5)
.br

.br
last\ modified:\ Fri\ Jul\ 15\ 15:42:38\ JST\ 2022
.br
Contact\ us:\ uecinfo@usp-lab.com
.br
Copyright\ (c)\ 2012-2022\ Universal\ Shell\ Programming\ Laboratory\ All\ Rights
.br


COMMANDS.HS/join2.hs
#!/usr/bin/env runghc --
import System.Environment
import System.IO
import Text.ParserCombinators.Parsec
import Control.Monad
import Data.ByteString.Lazy.Char8 as BS hiding (take,drop,filter,head,last,map,zip,repeat)
import Control.Applicative hiding ((<|>), many)
import Data.Char

{--
join2（Open usp Tukubai）

designed by Nobuaki Tounaka
written  by Hinata Yanagi

The MIT License

Copyright (C) 2022 Universal Shell Programming Laboratory

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
--}

showUsage :: IO ()
showUsage = do
    System.IO.hPutStr stderr "Usage    : join2 [+ng] <key=n> <master> [<tran>]\n"
    System.IO.hPutStr stderr "Version  : Sat Oct  1 16:50:42 JST 2022\n"
    System.IO.hPutStr stderr "Open usp Tukubai (LINUX+FREEBSD)\n"

main :: IO ()
main = do args <- getArgs
          case args of
                []                      -> showUsage
                ["-h"]                  -> showUsage
                ["--help"]              -> showUsage
                ["--version"]           -> showUsage
                [delim,key,master,tran] -> mainProc delim key master tran
                [a,b,c]                 -> if (a !! 0) == 'k' 
                                           then mainProc "" a b c
                                           else mainProc a b c "-"
                [key,master]            -> mainProc "" key master "-"
                _                       -> showUsage


mainProc :: String -> String -> String -> String -> IO ()
mainProc delim key master tran = do ms <- readF master
                                    ts <- readF tran
                                    mainProc' (parseKey key) ms ts (readDelim delim)

readF :: String -> IO BS.ByteString
readF "-" = BS.getContents
readF f   = BS.readFile f

readDelim :: String -> String
readDelim ""            = ""
readDelim ('+':str)     = str
readDelim ('-':'d':str) = str

parseKey :: String -> Keys
parseKey str = case parse keys "" str of
                    Right opt -> opt
                    Left  err -> Error (show err)

mainProc' :: Keys -> BS.ByteString -> BS.ByteString -> String -> IO ()
mainProc' (Keys ks) ms ts delim = out (join2 (makeDummy mlines delim) (head mlines) (drop 1 mlines) tlines)
                               where mlines = parseMaster ks (BS.lines ms) delim
                                     tlines = parseTran ks (BS.lines ts)

out :: [OutTran] -> IO ()
out []               = do return ()
out (ln:lns) = (BS.putStrLn ln) >> (out lns)

join2 :: [BS.ByteString] -> Master -> [Master] -> [Tran] -> [OutTran]
join2 _ _  _ []      = []
join2 dv (Master mk v) [] ((Tran p tk a):ts)
  | mk == tk      = makeLine (Master mk v) (Tran p tk a) : join2 dv (Master mk v) [] ts
  | otherwise     = makeLine (Master tk dv) (Tran p tk a) : join2 dv (Master mk v) [] ts
join2 dv (Master mk v) (m:ms) ((Tran p tk a):ts)
  | mk == tk      = makeLine (Master mk v) (Tran p tk a) : join2 dv (Master mk v) (m:ms) ts
  | mk <  tk      = join2 dv m ms ((Tran p tk a):ts)
  | otherwise     = makeLine (Master tk dv) (Tran p tk a) : join2 dv (Master mk v) (m:ms) ts

{--
join2 :: [Master] -> Tran -> OutTran
join2 ms (Tran p k a) = makeLine (pickMaster ms k) (Tran p k a)
--}

makeLine :: Master -> Tran -> OutTran
makeLine (Master [] v) (Tran p k a) = BS.unwords $ p ++ k ++ v ++ a
makeLine (Master k v) (Tran p _ a) = BS.unwords $ p ++ k ++ v ++ a

myWords :: BS.ByteString -> [BS.ByteString]
myWords line = BS.split ' ' line

pickMaster :: [Master] -> [BS.ByteString] -> Master
pickMaster ms k = if Prelude.length matched > 0 then head matched else dummy
            where matched = filter ( matchMaster k ) ms
                  a = last ms
                  dummy = f a
                  f (Master _ v) = Master [] v

matchMaster k (Master a b) = k == a
              
parseMaster :: [Int] -> [BS.ByteString] -> String -> [Master]
parseMaster ks lines delim = [ f (Prelude.length ks) (myWords ln) | ln <- lines ]
                             where f n ws = Master (take n ws) (drop n ws)

makeDummy :: [Master] -> String -> [BS.ByteString]
makeDummy ms "" = [ BS.pack $ take y ( repeat '_' ) | y <- x ]
               where x = maxLengths [ getValueLength m | m <- ms ]
makeDummy ms str = [ BS.pack str | y <- (g (head ms)) ]
               where g (Master _ b) = b

maxLengths :: [[Int]] -> [Int]
maxLengths [] = []
maxLengths [vsr] = map (\_ -> 1) vsr
maxLengths (vsr:vsl:[]) = [ if (fst z) > (snd z) then fst z else snd z | z <- (zip vsr vsl)] 
maxLengths (vsr:vsl:vss) = maxLengths (a:vss)
                            where zs = zip vsr vsl
                                  a  = [ if (fst z) > (snd z) then fst z else snd z | z <- zs] 

getValueLength :: Master -> [Int]
getValueLength (Master k v) = map ( wc . BS.unpack ) v

parseTran :: [Int] -> [BS.ByteString] -> [Tran]
parseTran ks lines = [ parseTran' ks (myWords ln) | ln <- lines ]

parseTran' :: [Int] -> [BS.ByteString] -> Tran
parseTran' ks ws = Tran (take pre ws) (take (Prelude.length ks) rem) (drop (Prelude.length ks) rem)
             where pre = (ks !! 0) - 1
                   rem = drop pre ws

wc :: String -> Int
wc []           = 0
wc (c:[])       = 1
wc (c:a:[])     = 2
wc (c:a:b:cs) = wc' $ ord c
                 where wc' n = if n < 128 
                       then 1 + wc (a:b:cs)
                       else (hanzen (n*256*256+(ord a)*256+(ord b))) + wc cs
                       hanzen m = if m >= 0xEFBDA1 && m <= 0xEFBE9F then 1 else 2

data Keys   = Keys [Int] | Error String
data Master = Master [BS.ByteString] [BS.ByteString] deriving Show -- keys and values
data Tran   = Tran [BS.ByteString] [BS.ByteString] [BS.ByteString] deriving Show -- values and keys and values
type OutTran = BS.ByteString

keys = string "key=" >> (try(rangekey) <|> try(singlekey) )

singlekey = do n <- many1 digit
               return $ Keys [read n::Int]

rangekey = do n <- many1 digit
              char '/'
              m <- many1 digit
              return $ Keys [(read n::Int)..(read m::Int)]
